### DISCUSSION TITLE
Windows platform protocol activation

### DISCUSSION THREAD

**Discussion by RedChops** [dotnet/maui#6493] [Q&A]
I'm a bit confused as to how we're supposed to use protocols in Windows applications. Let's say I have a protocol registered as 'myapp' and call 'myapp://?someparam=somevalue' (or 'myapp:?someparam=somevalue', or 'myapp://path?...', I've tried all three). In the past, I was able to override OnActivated in App.xaml.cs and pull the arguments out of the URI following steps like https://docs.microsoft.com/en-us/windows/uwp/launch-resume/handle-uri-activation.

With MAUI, there is no OnActivated to override in App.xaml.cs. Clicking a protocol link always just opens a new app window. If I override OnLaunched there the 'Arguments' property of LaunchActivatedEventArgs is always empty. If I try to create an OnActivated event handler using the new ConfigureLifecycleEvents method on MauiApp I get an empty Arguments property there too.

So basically two questions: How do I stop protocol activation from opening a new window, and how do I actually get the protocol arguments out of a protocol activation?

**Comment by GioviQ** [comment]
Have you try https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/communication/authentication?view=net-maui-7.0 ?

**Reply by RedChops** [reply]
That's different though, that's for WebAuthenticator which I am not using, and doesn't say anything about protocol links opening a second, separate application

**Comment by mstefarov** [comment]
I just ran into the same problem, and eventually found a couple workarounds.

One way is to subscribe to `Microsoft.Windows.AppLifecycle.AppInstance.GetCurrent().Activated` event from the ModuleInitializer.  From the callback, you can check if args.Data can be cast to IProtocolActivatedEventArgs, and get the Uri from there.

Another way is to override `OnLaunched` in your MauiWinUIApplication (App.xaml.cs).  Confusingly, the args object passed here is just a placeholder, and you have to get the real one by calling `Microsoft.Windows.AppLifecycle.AppInstance.GetCurrent().GetActivatedEventArgs()`.

**Reply by datvm** [reply]
This was the main issue I had and even after getting the relevant information from your comment, it's still frustrating because another instance took over instead. Thankfully I could solve it now.

**Comment by OnepunchmanSven** [comment]
@mstefarov how can I have the uri to activate the existing window? I can get the tokens back from the uri, but it's not an action like activation, the app still minimized there. And by using the protocol there, the only way is to have the returned uri like:
myapp://auth?token=xxxxxx

but usually the uri we get from the website is:
https://identity.getpostman.com/browser-auth/success?redirect_uri=postman://auth/callback&code=xxxxxxxx

I'm listing the postman example for your reference, this is not working for my app.

**Comment by Jockeyvb** [comment]
@RedChops 

my vb6 reg program to system Registry code 
try convert .cs ,may be ok:
if ProtocolHead  value  "myapp" well can call  you app in windows
 `Public Function RegProtocol(ProtocolHead As String, ExeFilePath As String, Optional AvgMax As Integer = 1, Optional ComputerName As String) As Boolean
Dim s As String, sAvg As String, i%
Dim TorF As Boolean

DeleteKey eHKEY_CLASSES_ROOT, ProtocolHead, ComputerName
 CreateKey eHKEY_CLASSES_ROOT, ProtocolHead & "\shell\open\command", ComputerName
SetDefaultValue eHKEY_CLASSES_ROOT, ProtocolHead, ProtocolHead & " Protocol", ComputerName
SetValue eHKEY_CLASSES_ROOT, ProtocolHead, "URL Protocol", ExeFilePath, ComputerName
 CreateKey eHKEY_CLASSES_ROOT, ProtocolHead & "\DefaultIcon", ComputerName
SetDefaultValue eHKEY_CLASSES_ROOT, ProtocolHead & "\DefaultIcon", ExeFilePath & ",1", ComputerName
For i = 1 To AvgMax
    sAvg = " " & Chr(34) & sAvg & "%" & i & Chr(34)
Next

 

TorF = SetDefaultValue(eHKEY_CLASSES_ROOT, ProtocolHead & "\shell\open\command", Chr(34) & ExeFilePath & Chr(34) & sAvg, ComputerName)
 RegProtocol = TorF
 Exit Function

End Function`
 

`RegProtocol "myapp", AppPath & App.EXEName & ".exe" `

**Comment by datvm** [comment]
I finally solved this issue, see https://stackoverflow.com/a/79123126/653457
